### DISCUSSION TITLE
Using `google` provider impersonation when running Terratest.

### DISCUSSION THREAD

**Discussion by jacobstr** [gruntwork-io/knowledge-base#776] [Help]
I'd like to use service account impersonation when running terratest. Ideally, I could inject these options in the `terraform.Options` but right now, I find myself using the `GOOGLE_IMPERSONATE_SERVICE_ACCOUNT` environment variable. This ... actually works great but I don't like the side-effecty nature of calling an `os.Setenv` in a helper function and combining that call with the `WithDevelopmentOptions` helper documented below.

For context, my goal is to set up a helper method like:

```
func WithDevelopmentOptions(t testing.TestingT, opts *terraform.Options) *terraform.Options {
...
}
```

To configure sane, safe development defaults to provide more paranoid controls. I want to make it easy for engineers to bootstrap their test harness without the footgun of running terratest on whatever `google_application_credentials` happens to be set to. The service account we impersonate in testing would be isolated to a specific GCP project.

---

<ins datetime="2023-10-17T22:53:32Z">
  <p><a href="https://support.gruntwork.io/hc/requests/110539">Tracked in ticket #110539</a></p>
</ins>

**Comment by denis256** [comment]
Hi,
was considered to use `t.Setenv()`?

from `testing.T.Setenv` documentation:

> Setenv calls os.Setenv(key, value) and uses Cleanup to restore the environment variable to its original value after the test.
